### PR TITLE
Port of the last 80X/90X MET (#16174 #17524) changes in 91X

### PR DESCRIPTION
**Pull request by mmarionncern**
Contains the fixed content of #17720 (similar to #17524) merged on top of #17801.

This PR ports the last MET changes developed in 80X, including the muon removal tools developed for Moriond 2017.

Additions :
- bad muon removal code for MET
- EGamma tools for MET
- 2016 legacy MET filters (remastered in the same way as in #17968 )

Limited matrix tests successful, and the use of the command 
cmsDriver.py step4  --conditions auto:run2_data_relval -s PAT,DQM:@miniAODDQM \
--runUnscheduled  --process PAT --data  --era Run2_2016 --eventcontent MINIAOD,DQM\
 --hltProcess reHLT --scenario pp --datatier MINIAOD,DQM \
--customise_unsch PhysicsTools/PatAlgos/slimming/customizeMiniAOD_MuEGFixMoriond2017.customizeAll\
 -n -1 --filein file:step3.root  --fileout file:step4.root --no_exec

with the step3.root file from 136.761 works as well.

This PR replaces #17720

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by Dr15Jones** [comment]
@Dr15Jones

File: PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py [hunk 1]
```diff
@@ -12,105 +12,116 @@ def applySubstructure( process ) :
...
-    task.add(process.ak8PFJetsCHSSoftDropMass)
-    process.patJetsAK8.userData.userFloats.src += ['ak8PFJetsCHSPrunedMass','ak8PFJetsCHSSoftDropMass']  
-    process.patJetsAK8.addTagInfos = cms.bool(False)
+    addToProcessAndTask('ak8PFJetsCHSPruned'+postfix, ak8PFJetsCHSPruned.clone(), process, task)
+    addToProcessAndTask('ak8PFJetsCHSSoftDrop'+postfix, ak8PFJetsCHSSoftDrop.clone(), process, task)
+    #process.load("RecoJets.JetProducers.ak8PFJetsCHS_groomingValueMaps_cfi")
+    #task.add(process.ak8PFJetsCHSPrunedMass)
+    #task.add(process.ak8PFJetsCHSTrimmedMass)
+    #task.add(process.ak8PFJetsCHSFilteredMass)
+    #task.add(process.ak8PFJetsCHSSoftDropMass)
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py]: is this commented out code needed?
Please remove or add comments inline in the code why the commented out block is relevant

File: PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py [hunk 2]
```diff
@@ -12,105 +12,116 @@ def applySubstructure( process ) :
...
     #add AK8 from PUPPI
-    process.load('RecoJets.JetProducers.ak8PFJetsPuppi_cfi')
-    task.add(process.ak4PFJetsPuppi)
-    task.add(process.ak8PFJetsPuppi)
+    #process.load('RecoJets.JetProducers.ak8PFJetsPuppi_cfi')
+    from RecoJets.JetProducers.ak8PFJetsPuppi_cfi import ak4PFJetsPuppi, ak8PFJetsPuppi
+    addToProcessAndTask('ak4PFJetsPuppi'+postfix,ak4PFJetsPuppi.clone(), process, task)
+    addToProcessAndTask('ak8PFJetsPuppi'+postfix,ak8PFJetsPuppi.clone(), process, task)
+    #task.add(process.ak4PFJetsPuppi)
+    #task.add(process.ak8PFJetsPuppi)
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/applySubstructure_cff.py]: here and in a few more places in this file it looks like the new commented out lines are not needed (they are replaced by a more appropriate load of modules/tasks

File: PhysicsTools/PatAlgos/python/slimming/customizeMiniAOD_MuEGFixMoriond2017.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        task.add(process.pfEGammaToCandidateRemapperBeforeGSFix)
+        process.reducedEgammaBeforeGSFix.gsfElectronsPFValMap = cms.InputTag("pfEGammaToCandidateRemapperBeforeGSFix","electrons")
+        process.reducedEgammaBeforeGSFix.photonsPFValMap      = cms.InputTag("pfEGammaToCandidateRemapperBeforeGSFix","photons")
+        # then the new one
+        process.pfEGammaToCandidateRemapper.electrons = cms.InputTag("gedGsfElectronsFixed")
+        process.pfEGammaToCandidateRemapper.photons   = cms.InputTag("gedPhotonsFixed")
+        process.pfEGammaToCandidateRemapper.electron2pf = cms.InputTag("particleBasedIsolationGSFixed","gedGsfElectrons")
+        process.pfEGammaToCandidateRemapper.photon2pf   = cms.InputTag("particleBasedIsolationGSFixed","gedPhotons")
+    else:
+        print "WARNING : attempt to use gain switch corrected electron/photon collection gedGsfElectronsFixed, but the current process does not contain such collection"
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/customizeMiniAOD_MuEGFixMoriond2017.py]: please change to sys.stderr.write(

**Comment by mmarionncern** [comment]
code cleaned

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
after 5 hours jenkins job for comparison is still pending 👎

**Comment by slava77** [comment]
Looking at jenkins comparisons I see that MC DQM has puppi jet area disappear with this PR (red is with this PR)
e.g. 10024:
wf10024_slimmedjetpuppi_area

It doesn't looks like a bugfix.

@mmarionncern please update the PR description to include the features introduced in this PR (links to old closed PRs are not satisfactory).

**Comment by slava77** [comment]
the puppi jet area issue was not present in an older version of #17524 (I have a snapshot of jenkins results as of Mar 10). Either the remaining changes made in #17524  https://github.com/cms-sw/cmssw/compare/d55e6aae4b73d90f982a3c18b2c28a6a95c6c686...ba8f8656335aa95e2d58d66ff9fdaba1d63f70b4 have introduced the issue or this came with the additional changes made on top of #17524

**Comment by mmarionncern** [comment]
@slava77 puppi jet area issue fixed (one line was in an unwanted if condition) and description added

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/18165	1f98762
- changes are in line with the description
- jenkins tests pass and comparisons show no relevant differences
- local tests show expected behavior:
    - in wf 25202 the bad particle filters show up in the workflow execution; their CPU cost is negligible
    - wf 136.761 output in AOD can be processed with miniAOD (PAT step) and PhysicsTools/PatAlgos/slimming/customizeMiniAOD_MuEGFixMoriond2017.customizeAll (the workflow runs to completion on 200 events)

**Comment by monttj** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**